Transport shutdown timing fix

was4 opened the pull request:
* Deconflict lock flags
 * Debug negative references
 * Transport shutdown timing fix

Commit:
Deconflict lock flags

Restore missing patch (or some variant).

Pass lock flags to svc_rqst_xprt_unregister() for future use.

Bugfix: missing condition on dplx_rec unlock.

Bugfix: clear from svc_xprt table sooner to reduce MT-thread window
possibly re-using SVCXPRT (although checking _DESTROYED should prevent).

Signed-off-by: William Allen Simpson <william.allen.simpson@redhat.com>

Commit:
Debug negative references

Log negative cc_refs, cl_refs, ev_refs, xp_refs; serendipitously
matching existing svc_xprt and svc_rdma debugging.

Bugfix: use atomic_fetch for ev_refs check in lookup.

Signed-off-by: William Allen Simpson <william.allen.simpson@redhat.com>

Commit:
Transport shutdown timing fix

Partial revert commit 3953171e833431463cdeec4b806eca82d7fe43c0
    Transport memory free task

The thread scheduler doesn't necessarily do work in the order submitted.
In this case, the scheduler can/will run the per transport cleanup task
before other ready worker tasks, causing timing conflicts with the
service request channel task. Use the channel task to free transports.

Client call timeout is most likely increasing relative to other calls
for the same connection, worst case for rbtree insertion. Profiling
showed that rbtree was not a good choice, and TAILQ improved by 200%.

Moreover, maintaining the potentially large list of timeouts in the
service request was slowing event response. Instead, keep only one
entry per transport, with callback timeouts maintained by transport.

Signed-off-by: William Allen Simpson <william.allen.simpson@redhat.com>

dang reviewed (commented):
Okay, so this doesn't fix the issue.  I have an easy reproducer now (for some definition of easy) and it still happens with this applied.

I'll start instrumenting, and we can discuss tomorrow.

mattbenjamin commented on ntirpc/rpc/clnt.h:
```diff
@@ -114,7 +114,7 @@ typedef struct rpc_client {
 
 	mutex_t cl_lock;	/* serialize private data */
 	struct rpc_err cl_error; /* specific error code */
-	uint32_t cl_refcnt;	/* handle refcnt */
```
what does changing refcnt to refs do?  I would request that we stay consistent with other uses of refcnt in the ntirpc and nfs-ganesha source

mattbenjamin commented on src/clnt_internal.h:
```diff
@@ -46,7 +46,7 @@ static inline void
 clnt_data_init(struct cx_data *cx)
 {
 	mutex_init(&cx->cx_c.cl_lock, NULL);
-	cx->cx_c.cl_refcnt = 1;
```
ditto

mattbenjamin reviewed (commented):
what was being being inserted into an rbtree on a per call basis?  this does not appear to be the reply matchin case?

that said "Client call timeout is most likely increasing relative to other calls
for the same connection, worst case for rbtree insertion. Profiling
showed that rbtree was not a good choice, and TAILQ improved by 200%" seems suspicious.  why do we need to update a timer list on every call?

was4 commented:
On 4/25/18 12:47 PM, Daniel Gryniewicz wrote:
> *@dang* commented on this pull request.
> 
> Okay, so this doesn't fix the issue. I have an easy reproducer now (for some definition of easy) and it still happens with this applied.
> 
> I'll start instrumenting, and we can discuss tomorrow.
> 
Good thing I checked email before my nap.  I'd just pushed an
update patch.

I've found one fairly old issue so far:

-                       rpc_dplx_rui(rec);
+                       if (!(flags & RPC_DPLX_LOCKED))
+                               rpc_dplx_rui(rec);

So in that case (should be rare) it was unilaterally unlocked, even
though it wasn't supposed to be.

Found it because I was resolving some symbol overlap a bit earlier:

-       if (!(flags & SVC_RQST_FLAG_LOCKED))
+       if (!(flags & RPC_DPLX_LOCKED))

****_FLAG_LOCKED should be the xp_lock (0x00010000), not the
wrapping rpc_dplx_rec lock.  I've updated those symbols (0x00100000),
as well as an internal SVC_RQST_LOCKED (0x01000000) for good measure.

I remember fixing that sometime in the past year.  But it seems that
I'd accidentally dropped the patch somehow....

Anyway, I've also found that timeouts are sometimes going backward,
so I've given up on compatibility and changed fields to timespec.

And still have a refcounting problem somewhere.

was4 commented:
Likewise, this was another was16backport.  It fixed my crash
under heavy load.  Has this made it into the tree?
